packages/wallet/api/src/sharingUnserializer.js

```diff
@@ -0,0 +1,81 @@
+// @ts-check
+import { Far, makeMarshal } from '@endo/marshal';
+
+/**
+ * Make a pair of unserializers that share low-privilege values.
+ *
+ * The "high" privilege unserializer can understand objects as slots, such as
+ * those from the wallet serializer. The "low" privilege unserializer can only
+ * understand strings as slots, such as those from the board serializer.
+ *
+ * This provides a way, for instance, to get identical brand references from
+ * both a public contract and a user's purses.
+ *
+ * @param {string} [lowProperty] - The key name to check on a high privilege
+ * slot object. If the string value at this key is equivalent to the value of
+ * another low privilege slot, the high privilege unserializer will return the
+ * same object for both slots.
+ */
+export const makeSharingUnserializer = lowProperty => {
+  assert(lowProperty, 'lowProperty must be specified for sharing unserializer');
+  const seen = {
+    // high: new WeakMap(),  // ISSUE: WeakMap conflicts with string keys
+    high: new Map(),
+    low: new Map(),
+  };
+
+  const slotToVal = {
+    /**
+     * @param {Record<string, any>} slot
+     * @param {string} iface
+     */
+    high: (slot, iface) => {
+      let obj;
+      let key;
+
+      if (lowProperty in slot) {
+        key = slot[lowProperty];
+        if (seen.low.has(key)) {
+          return seen.low.get(key);
+        }
+        obj = Far(iface, {});
+        seen.low.set(key, obj);
+      } else {
+        key = JSON.stringify(slot);
+        if (seen.high.has(key)) {
+          return seen.high.get(key);
+        }
+        obj = Far(iface, {});
+        seen.high.set(key, obj);
+      }
+      return obj;
+    },
+
+    /**
+     * @param {Record<string, any>} slot
+     * @param {string} iface
+     */
+    low: (slot, iface) => {
+      if (seen.low.has(slot)) {
```

Comment:
`Map` (and our `scalarMapStore`) keys off `Object` identity. If `slot` is an `Object` and not a string, a caller who sends `{ foo, bar }` first and then (a distinct) `{ foo, bar}` second, will get two different presences, when they should get the same. The supplier of the serialized data may or may not supply those semantically-identical slot values as identical objects, it depends upon how they're getting them. If they're starting with strings, and doing a `JSON.parse`, then they're almost certainly going to be distinct values, so you'll get unique `Far` presences for each act of deserialization.

I'd highly recommend using string keys if you can find a way to do it, to avoid both of these problems.

And you should definitely have a test that deserializes the same data twice and asserts that you're getting identical (`test.is(old, new)`) presences.